Add examples to CodeService.md docs (#6050)
iTwin/itwinjs-core: docs/learning/backend/CodeService.md

- `reserveCode` to reserve a single code
- `reserveCodes` to reserve a set of codes

Here are the steps to follow.

> The identity of the code service and information about how to connect are stored in the iModel persistently. When you open a BriefcaseDb for that iModel, it will read the configuration data from the iModel and automatically try to connect to the code service. When an app opens a briefcase, the briefcase will connect to the code service using the connection information stored in the iModel. The briefcase `.codeService` property will refer to the connected code service client, and apps must use that to reserve, update, and query codes and CodeSpecs.

Reviewer: ```suggestion
The identity of the code service and information about how to connect are stored in the iModel persistently. When you open a BriefcaseDb for that iModel, it will read the configuration data from the iModel and automatically try to connect to the code service. When an app opens a briefcase for editing, the briefcase will connect to the code service using the connection information stored in the iModel and the user's credentials. The briefcase `.codeService` property will refer to the connected CodeService client, and apps must use that to reserve, update, and query codes and CodeSpecs.
```

Author: done

Here is an example of using codeService to reserve a new internal code.

``` ts
const props = await BriefcaseManager.downloadBriefcase({ accessToken, iTwinId, iModelId, asOf: { afterChangeSetId }, briefcaseId: 0, fileName });
const iModel = await BriefcaseDb.open({ fileName: props.fileName });
```

``` ts
[[include:CodeService.reserveInternalCodeForNewElement]]
```

The logic to reserve an external code is similar - just use the `externalCodes` property instead.

> Note that reserveCode is asynchronous. It locks and unlocks the code index in the cloud.

Reviewer: ```suggestion
Note that reserveCode is asynchronous and must be `await`ed. It locks and then updates the code index in the cloud.
```

Author: done

Note the use of the `CodeService.makeProposedCode` helper function.

> Note that the reserveCode function will throw if the code is already reserved. If that happens when you are trying to insert a new element, it means that you cannot use that code.

Reviewer: ```suggestion
The `reserveCode` function will throw if the code is already reserved. If that happens when you are trying to insert a new element, it means that someone else has already reserved that code and you should choose a different one.
```

Author: done

Don't call reserveCode when updating an existing element. If the element's code is changing, call `updateCode` before updating it.

In general, to find out if a code has been reserved, call findCode.

```ts
[[include:CodeService.findCode]]
```

There are also `CodeService` apis for updating properties of and deleting codes. They have the same rules for authentication and serialization as reserving Codes. They are:

- `updateCode` to modify the properties of a code
- `updateCodes` to modify the properties of a set of codes
- `deleteCodes` to delete a set of codes

For example,

``` ts
[[include:CodeService.updateInternalCodeForExistinglement]]
```

> Note that the element's `federationGuid` must be used as the Code's GUID.

Reviewer: ```suggestion
Note that the element's `federationGuid` is the Code's Guid.
```

Author: done

- `reserveCode` to reserve a single code
- `reserveCodes` to reserve a set of codes

Here are the steps to follow.

First, you must choose the code service that should be used at the time that you create your iModel. The identity of the code service and information about how to connect  are stored in the iModel persistently.

When you open a BriefcaseDb for that iModel, it will read the configuration data from the iModel and automatically try to connect to the code service. To support that, apps that access the iModel must provide the client for the code service and set `CodeService.createForIModel` to point to it.

For example, if you configure your new iModel to use Bentley's cloud-based code service, apps must provide the `ITwinCodeService` client. To do that, an app simply depends on the `@bentley/code-service` package and then imports it.

```ts

> import { ITwinCodeService } from "@bentley/code-service";

Reviewer 2: This is currently an internal package so can't be in our public facing docs. Can you take out the section talking about this as an example?

Author: I made the requested changes

Reviewer: > How will the apps that our customers write deal with internal codes?

They will use iTS.
